### PR TITLE
ApplicationInsights performance improvements

### PR DESCRIPTION
**Pull request by mathewc**
A set of performance improvements for the AI logging pipeline. Here are the averaged Loader.io run results for a 1 minute single instance P2V2 Hello World scenario:

|| # Requests | Latency (ms) |
|-|------|---------|
|Before| 65,957 |90 |
|After| 69,519 | 86 |

So about a **5% improvement** in RPS/Latency.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs [hunk 1]
```diff
@@ -32,7 +34,7 @@ internal class ApplicationInsightsLogger : ILogger
         internal const string MetricMaxKey = "max";
         internal const string MetricStandardDeviationKey = "standarddeviation";
 
-        private static readonly string[] SystemScopeKeys =
+        private static readonly HashSet<string> SystemScopeKeys = new HashSet<string>
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:37]: For a single function execution we were searching this array for about 25 scope properties.

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs [hunk 2]
```diff
@@ -193,22 +192,18 @@ private void LogException(LogLevel logLevel, IEnumerable<KeyValuePair<string, ob
...
 
-            ApplyScopeAndStateProperties(telemetry, values);
+            ApplyScopeAndStateProperties(telemetry.Properties, values);
 
             _telemetryClient.TrackException(telemetry);
         }
 
         private void LogTrace(LogLevel logLevel, IEnumerable<KeyValuePair<string, object>> values, string formattedMessage)
         {
-            TraceTelemetry telemetry = new TraceTelemetry(formattedMessage)
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:203]: Many of the signature changes above were to allow me to remove this object creation, and instead use the overload below that takes the values directly.

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs [hunk 3]
```diff
@@ -264,20 +259,24 @@ private static void ApplyProperty(ISupportProperties telemetry, string key, obje
                 stringValue = value.ToString();
             }
 
-            telemetry.Properties[$"{propertyPrefix}{key}"] = stringValue;
+            string prefixedKey = applyPrefix ? _prefixedProperyNames.GetOrAdd(key, k =>
+            {
+                return $"{LogConstants.CustomPropertyPrefix}{k}";
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:264]: Saving a ton of string allocations by caching these

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/ApplicationInsightsServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -70,7 +70,6 @@ public static IServiceCollection AddApplicationInsights(this IServiceCollection
             });
 
             services.AddSingleton<ITelemetryInitializer, WebJobsRoleEnvironmentTelemetryInitializer>();
-            services.AddSingleton<ITelemetryInitializer, WebJobsSanitizingInitializer>();
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/ApplicationInsightsServiceCollectionExtensions.cs:73]: Similar to performance work I did in the Functions Host system logging, we don't want to be running sanitizer on every single piece of data. Instead, we should run any required sanitization at the call site for any data we think might be sensitive.

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsRoleEnvironmentTelmetryInitializer.cs [hunk 5]
```diff
@@ -32,34 +35,27 @@ public void Initialize(ITelemetry telemetry)
...
             }
 
-            // We cannot cache these values as the environment variables can change on the fly.
-            if (string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+            Lazy<string> siteSlotName = new Lazy<string>(() =>
             {
-                telemetry.Context.Cloud.RoleName = GetRoleName();
-            }
+                // We cannot cache these values as the environment variables can change on the fly.
+                return GetAzureWebsiteUniqueSlotName();
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsRoleEnvironmentTelmetryInitializer.cs:41]: Previously we were looking up env values and formatting the site name multiple times for each log.

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsRoleEnvironmentTelmetryInitializer.cs [hunk 6]
```diff
@@ -32,34 +35,27 @@ public void Initialize(ITelemetry telemetry)
...
+            var internalContext = telemetry.Context.GetInternalContext();
+            if (string.IsNullOrEmpty(internalContext.NodeName) &&
+                !string.IsNullOrEmpty(siteSlotName.Value))
             {
-                // maintain previous behavior of node having the full url
-                name = name + WebAppSuffix;
+                internalContext.NodeName = _siteNodeNames.GetOrAdd(siteSlotName.Value, p =>
+                {
+                    // maintain previous behavior of node having the full url
+                    return p += WebAppSuffix;
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsRoleEnvironmentTelmetryInitializer.cs:56]: Saving an allocation per log

File: src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Processors/FilteringTelemetryProcessor.cs [hunk 7]
```diff
@@ -49,7 +49,7 @@ private bool IsEnabled(ITelemetry item)
                 if (telemetry.Properties.TryGetValue(LogConstants.LogLevelKey, out string logLevelString) &&
                     Enum.TryParse(logLevelString, out LogLevel logLevel))
                 {
-                    LoggerFilterRule filterRule = _ruleMap.GetOrAdd(categoryName, SelectRule(categoryName));
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Processors/FilteringTelemetryProcessor.cs:52]: Simple mistake meaning we were running the rule selection logic for every log

**Review by brettsam — APPROVED** [review]

**Pull request merged**